### PR TITLE
[core] Re-activating DEM in CI

### PR DESCRIPTION
**Pull request by maceligueta**
It was disabled due to some error yielded by the CI. This PR is to remember why and try to fix it.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -106,6 +106,7 @@ jobs:
         set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%KRATOS_APP_DIR%\MeshMovingApplication;
         set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%KRATOS_APP_DIR%\EigenSolversApplication;
         set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%KRATOS_APP_DIR%\StructuralMechanicsApplication;
+        set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%KRATOS_APP_DIR%\DEMApplication;
```
Comment (philbucher) [inline .github/workflows/ci.yml:110]: this is the necessary change, not sure what the others do

File: scripts/build/dist/configure_ubuntu_14_04.sh [hunk 2]
```diff
@@ -29,6 +29,7 @@ export PYTHON_EXECUTABLE="/usr/bin/python3.5"
 add_app ${KRATOS_APP_DIR}/ExternalSolversApplication
 add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
 add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication
+add_app ${KRATOS_APP_DIR}/DEMApplication
```
Comment (philbucher) [inline scripts/build/dist/configure_ubuntu_14_04.sh:32]: don't know what this does

File: scripts/build/dist/configure_ubuntu_16_04.sh [hunk 3]
```diff
@@ -29,6 +29,7 @@ export PYTHON_EXECUTABLE="/usr/bin/python3.5"
 add_app ${KRATOS_APP_DIR}/ExternalSolversApplication
 add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
 add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication
+add_app ${KRATOS_APP_DIR}/DEMApplication
```
Comment (philbucher) [inline scripts/build/dist/configure_ubuntu_16_04.sh:32]: same

**Review by philbucher — COMMENTED** [review]
Thanks!

File: scripts/build/dist/configure_ubuntu_14_04.sh (hunk 2 above)
Comment (maceligueta) [reply scripts/build/dist/configure_ubuntu_14_04.sh:32]: I am happier if DEM is also here :-)

**Comment by philbucher** [comment]
Now it was able to run the tests (56s/60s) but there is also a failing test:
~~~
======================================================================
ERROR: test_DEM2D_inlet (test_DEM_2D_inlet.TestDEM2DInlet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "d:\a\Kratos\Kratos\bin\Custom\applications\DEMApplication\tests\test_DEM_2D_inlet.py", line 86, in test_DEM2D_inlet
    CreateAndRunStageInOneOpenMPThread(DEM2D_InletTestSolution, model, parameters_file_name)
  File "d:\a\Kratos\Kratos\bin\Custom\applications\DEMApplication\tests\test_DEM_2D_inlet.py", line 29, in CreateAndRunStageInOneOpenMPThread
    my_obj(model, project_parameters).Run()
  File "d:\a\Kratos\Kratos\bin\Custom\KratosMultiphysics\analysis_stage.py", line 50, in Run
    self.RunSolutionLoop()
  File "d:\a\Kratos\Kratos\bin\Custom\KratosMultiphysics\analysis_stage.py", line 69, in RunSolutionLoop
    self.FinalizeSolutionStep()
  File "d:\a\Kratos\Kratos\bin\Custom\applications\DEMApplication\tests\test_DEM_2D_inlet.py", line 56, in FinalizeSolutionStep
    self.CheckVel(node_vel, expected_value, tolerance)
  File "d:\a\Kratos\Kratos\bin\Custom\applications\DEMApplication\tests\test_DEM_2D_inlet.py", line 64, in CheckVel
    raise ValueError('Incorrect value for VELOCITY_Y: expected value was '+ str(expected_value) + ' but received ' + str(vel))
ValueError: Incorrect value for VELOCITY_Y: expected value was 0.38048924 but received -0.6035131139206471
~~~

Unfortunately we have no method yet to time the tests :/
Do you have a hunch or should we try them individually? Its only 17 luckily

**Comment by maceligueta** [comment]
I am compiling Kratos in my Windows laptop to see if that test also fails here. The compilation takes forever and it uses all processors, so I couldn't even use the web browser to answer you. lol

**Comment by maceligueta** [comment]
I have no idea why this test runs fine everywhere except here. Maybe some random number, which is created different in Windows than in Linux???
@farrufat-cimne

**Comment by farrufat-cimne** [comment]
> I have no idea why this test runs fine everywhere except here. Maybe some random number, which is created different in Windows than in Linux???
> @farrufat-cimne

It only fails on WIndows CI and the error is:
ValueError: Incorrect value for VELOCITY_Y: expected value was 0.38048924 but received -0.6035131139206471
So it is not some kind of out of tolerance issue.
I will check the script for some process that could insert randomness in the inlet injection (maybe a small degree in the random orientation at injection point?).

**Comment by maceligueta** [comment]
> I will check the script for some process that could insert randomness in the inlet injection (maybe a small degree in the random orientation at injection point?).
not in the orientation (it is set to zero). Maybe it's the injection point? It is chosen with a random, too.

**Comment by farrufat-cimne** [comment]
> > I will check the script for some process that could insert randomness in the inlet injection (maybe a small degree in the random orientation at injection point?).
> > not in the orientation (it is set to zero). Maybe it's the injection point? It is chosen with a random, too.

I don't think that could be the reason since this would probably be already detected in other Windows CI tests but, would it be posible that the Windows CI is assigning the node ID to the other sphere? (in here: if node.Id == 7:).
Since the test itself is pretty simple, just 2 injected particles impacting in almost normal trajectory (10% deviation)

**Comment by maceligueta** [comment]
> > > I will check the script for some process that could insert randomness in the inlet injection (maybe a small degree in the random orientation at injection point?).
> > > not in the orientation (it is set to zero). Maybe it's the injection point? It is chosen with a random, too.
> 
> I don't think that could be the reason since this would probably be already detected in other Windows CI tests but, would it be posible that the Windows CI is assigning the node ID to the other sphere? (in here: if node.Id == 7:).
> Since the test itself is pretty simple, just 2 injected particles impacting in almost normal trajectory (10% deviation)

There is no other Windows CI running tests. Appveyor is in Windows but it is not calling the tests.

**Comment by maceligueta** [comment]
Apparently, the error reproduces in my Windows laptop, too, with the same error exactly.
With @farrufat-cimne  we confirmed that it seems like the `rand()` function gives different values for Windows and Linux, so they the simulations do completely different things.
The options found on the web are:
- write my own rand()
- use the random generator from Boost

I think this may affect everyone and should be tackled as a new Kratos core feature.
What do you think @KratosMultiphysics/technical-committee?

**Comment by philbucher** [comment]
global question, why would you use a random number generator in a test?
To me sounds like it is doing its job, i.e. giving you a random number
I.e. assuming to get the same random number every times sounds contradictory

**Comment by maceligueta** [comment]
The generation of random numbers depends on a seed, which can be always the same or different for every run. Examples of seeds are: time step (always the same value for all simulations), wall clock (changes every time you run the case). In our case, we are interested in generating particles from random spots, but we want the code to generate the particles from the same spots every time we run it, so we can debug it. So we expect to have the same random sequence all times. This pseudo random sequence varies from Linux to Windows, but it is always the same if you stay in Linux, as we want to.

**Comment by philbucher** [comment]
Ok that is what I suspected, thx for clarifying

**Comment by philbucher** [comment]
@maceligueta how abt you hardcode the value for now such that we can add DEM back to the CI and discuss the issue you are raising in a separate issue?

**Comment by maceligueta** [comment]
Do you mean ‘if Windows... else...’ ?

**Comment by philbucher** [comment]
either this OR instead of using the randomly generated number hardcode the value
Whatever is easier

**Comment by philbucher** [comment]
looking at your code I guess this might even lead to different results on different operating systems?

**Comment by maceligueta** [comment]
I am not sure, in this specific case, the random which is affecting is is the one in this line.
https://github.com/KratosMultiphysics/Kratos/blob/5d33e770a09c2111fbc871e99c74a23ba14900f8/applications/DEMApplication/custom_utilities/inlet.cpp#L608
Apparently, rand() changes across platforms. But... I just found that there are other random generators which are totally portable, like this one:
http://www.cplusplus.com/reference/random/mt19937/
I will dedicate some time to it today and let's hope it works...

**Comment by philbucher** [comment]
Ok cool 

I thought I had seen this already in Kratos

@loumalouomega didnt you use random numbers somewhere?

**Comment by philbucher** [comment]
seems like @msandre used this in some tests:
https://github.com/KratosMultiphysics/Kratos/blob/5d33e770a09c2111fbc871e99c74a23ba14900f8/applications/StructuralMechanicsApplication/tests/cpp_advanced_constitutive_tests/test_total_lagrangian_element_matrices.cpp#L42-L44

**Comment by loumalouomega** [comment]
> @loumalouomega didnt you use random numbers somewhere?

Maybe...

**Comment by philbucher** [comment]
This one: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/random_initializer_utility.h

**Comment by maceligueta** [comment]
This presentation helped me out a bit: https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful
But it says that `uniform_int_distribution` is not invariant accross platforms, so maybe `uniform_real_distribution` isn't either, @philbucher @msandre

**Comment by maceligueta** [comment]
> This presentation helped me out a bit: https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful
> But it says that `uniform_int_distribution` is not invariant accross platforms, so maybe `uniform_real_distribution` isn't either, @philbucher @msandre

Exactly, this guy tested it and they are not portable: https://www.reddit.com/r/cpp/comments/30w7cs/inconsistency_in_c_random/

**Comment by loumalouomega** [comment]
> This one: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/random_initializer_utility.h

I don't remember why I did that 

imagen

**Comment by maceligueta** [comment]
Well it seems my last change did not fix anything, I keep debugging it both in Linux and Windows...

**Comment by maceligueta** [comment]
OK, I found something interesting, and anyone is welcome to help me out, please.
The problem was that the particles generated were created in a different order for Linux and Windows. @farrufat-cimne and me thought it was due to the random choice of the injection spot, but it wasn't. 
The actual problem is here:
https://github.com/KratosMultiphysics/Kratos/blob/5d33e770a09c2111fbc871e99c74a23ba14900f8/applications/DEMApplication/custom_utilities/inlet.cpp#L115
The iteration over the submodelparts is done in a different order. In Windows it is correct, it first looks to the first submodelpart. But in Linux it is wrong, it first goes to the second submodelpart, then goes to the first. Any clue of why this is happening?

**Comment by maceligueta** [comment]
Ok, speaking with @RiccardoRossi  we realized that the container of submodelparts is a std:unordered_map. The iterators on an unordered map are not standard. I will think of a different solution.

**Comment by philbucher** [comment]
you could get the names of the submodelparts, sort them an then iterate using them
This is done here:
https://github.com/KratosMultiphysics/Kratos/blob/5d33e770a09c2111fbc871e99c74a23ba14900f8/kratos/sources/model_part.cpp#L1802-L1811

**Comment by philbucher** [comment]
hm cool now it seems to work :)
If the restart tests take to long you could add them to the nightly suite instead

**Comment by maceligueta** [comment]
I saw your message about ordering the submp way after I pushed. I am not sure which version is better, mine with pointers or the one you linked, with names. I guess mine is a bit faster, but it will be an insignificant amount of time.

**Comment by maceligueta** [comment]
I think we can merge these modifications, if nobody is against.
Concerning the tests, I am surprised that they take that long in the CI. They take 5-10 seconds in our computers...!
We will discuss what to do with them. How to shorten them or how to prepare a new CI exclusively for tests (no compilation).

File: applications/DEMApplication/custom_utilities/inlet.cpp [hunk 4]
```diff
@@ -540,7 +552,7 @@ namespace Kratos {
 
             if (number_of_particles_to_insert) {
                 //randomizing mesh
-                srand(/*time(NULL)* */r_modelpart.GetProcessInfo()[TIME_STEPS]);
+                std::mt19937 random_generator(r_modelpart.GetProcessInfo()[TIME_STEPS]);
```
Comment (philbucher) [inline applications/DEMApplication/custom_utilities/inlet.cpp:555]: so you decided to keep this?

File: applications/DEMApplication/tests/test_DEMApplication.py [hunk 5]
```diff
@@ -57,8 +57,8 @@ def AssembleTestSuites():
 
     smallSuite.addTest(test_DEM_3D_continuum.TestDEM3DContinuum("test_DEM3D_continuum"))
 
-    smallSuite.addTest(test_restart.TestRestartOneBall("test_execution"))
-    smallSuite.addTest(test_restart.TestRestartTwoBalls("test_execution"))
+    #smallSuite.addTest(test_restart.TestRestartOneBall("test_execution"))
+    #smallSuite.addTest(test_restart.TestRestartTwoBalls("test_execution"))
 
     # Create a test suit with the selected tests plus all small tests
     nightSuite = suites['nightly']
```
Comment (philbucher) [inline applications/DEMApplication/tests/test_DEMApplication.py]:
```suggestion
    # Create a test suit with the selected tests plus all small tests
    nightSuite = suites['nightly']
    
    nightSuite.addTest(test_restart.TestRestartOneBall("test_execution"))
    nightSuite.addTest(test_restart.TestRestartTwoBalls("test_execution"))
```

**Review by philbucher — COMMENTED** [review]
@maceligueta to me this is fine, but I highly recommend adding the tests to the nightly

A few more comments:
- I took a few looks at your tests. 
    - I have the impression that you are running a complete simulation for each feature you are testing. In my experience this is not only cumbersome for debugging, i.e. sth starts failing but the error could be anywhere, but it also takes a lot of (testing-) time. 
    - The longer I write code, the smaller and more targeted my tests become, meanwhile I am writing many many small tests. I know this sounds like a lot of work (which it initially is) but I cannot count the times I thanked myself for doing this when I had to change / refactor things. Also test-driven development is no longer just a concept for me, I actively use it and think it made my developments much better overall and also a lot faster!
I would recommend you and @KratosMultiphysics/dem to take a look at the tests in e.g. CoSimulationApp, MappingApp, StructuralMechanics. I think meanwhile some quite nice testing has been established and proven to work.
    - Furthermore I recommend to exploit more the various test-suites we have (nightly and validation). The nightly is run every night, and has 15min (15 times more than the small-suite) of runtime. Even though this is not in the CI, it is still very helpful to know, when sth started failing (by checking the nightly logs).
    - Also C++ tests I would recommend for testing small features, e.g. the response of a Constitutive Law for a given set of input. This can of course be complementary to a longer testcase.
    - Remember that tests are only useful if they are being executed. I know from myself, that if some tests are taking too long I will just not run them, which defeats the purpose.
- Regarding the CI: 
    - Even though a CI without compiling it is ofc possible, I am not sure that we should do this:
        - What would be the benefit? More time for testing? Even now we could probably do a bit more in the CI (Maybe 2 instead of 1min). Consider that if every app exploits this soon we will have the same situation that we had with Appveyor and Travis, that we had to wait the entire day until the CI runs, which i think makes quite a big difference in working.
        - Which "base" would we use for running the tests? => if C++ code changes we HAVE TO recompile otherwise doesn't make sense. Hence I think it males no sense to use Docker here
        - Splitting the compilation from the execution of the tests (e.g. with Artifacts) could be done, but I am not convinced this is worth the effort
    - We already have the nightly tests that run longer (and in Release!) so I would recommend putting the longer tests there. This is e.g. what I do for the CoSimulation-tests, that inherently take long.
    - The current runtimes for the DEM tests are fine IMO, they differ a bit btw the compilers/OSs but if you say they run 5-10sec on your machines and 30-50 in the CI I would even say that the CI is quite fast! Usually my estimate is that the CI is ~10 times slower. Also consider that we are not using optimization flags for compiling since this makes the compilation time explode (see nightly build)

I hope this helps a bit.

also in the style of @roigcarlo: :D
**TL;DR** I recommend writing smaller tests and taking a look what other apps are doing

**Comment by maceligueta** [comment]
Hi @philbucher . I and the totality of the team agree with your comments, thanks for that. Actually, most of them we were already aware of.
After a couple of meetings in which we discussed about this, we realized that currently none of the members of the team is dedicating even a second to the maintenance of the DEM. Bugs are always max priority, this never changed, but refactoring and re-coding would require a dedication we lack at the moment.
We are trying to re-organize the team and see if the solution requires more staff or just switching tasks.

File: applications/DEMApplication/custom_utilities/inlet.cpp (hunk 4 above)
Comment (maceligueta) [reply applications/DEMApplication/custom_utilities/inlet.cpp:555]: Yes, I still think it is an improvement, though it should be done in many other places.

**Comment by maceligueta** [comment]
Btw, I am not getting any report from the nightly tests. I used to, and I do not remember unregistering from it. Is there any way to get an automatic report if the DEM nightly tests fail?

**Comment by philbucher** [comment]
> Hi @philbucher . I and the totality of the team agree with your comments, thanks for that. Actually, most of them we were already aware of.
> After a couple of meetings in which we discussed about this, we realized that currently none of the members of the team is dedicating even a second to the maintenance of the DEM. Bugs are always max priority, this never changed, but refactoring and re-coding would require a dedication we lack at the moment.
> We are trying to re-organize the team and see if the solution requires more staff or just switching tasks.

:+1:
I fully understand this! I am just recommending, at least you know how things could be improved :)

**Comment by philbucher** [comment]
> Btw, I am not getting any report from the nightly tests. I used to, and I do not remember unregistering from it. Is there any way to get an automatic report if the DEM nightly tests fail?

this has not yet been implemented. Since we switched to the github-CI, you can check the results in the `Actions` tab by filtering for `schedule` and then selecting one of the builds
image

@KratosMultiphysics/technical-committee I think we should discuss how we want to make the test-results accessible
maybe only send a mail to a team if tests are failing...?

**Review by philbucher — APPROVED** [review]
nice work, thanks for your help
